In [1]:
import numpy as np
import pandas as pd


In [31]:
house_ds = pd.read_csv("housing.csv")
house_ds.head(10)

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
0,7129300520,20141013T000000,221900,3,1.00,1180,5650,1.0,0,0,...,7,1180,0,1955,0,98178,47.5112,-122.257,1340,5650
1,6414100192,20141209T000000,538000,3,2.25,2570,7242,2.0,0,0,...,7,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639
2,5631500400,20150225T000000,180000,2,1.00,770,10000,1.0,0,0,...,6,770,0,1933,0,98028,47.7379,-122.233,2720,8062
3,2487200875,20141209T000000,604000,4,3.00,1960,5000,1.0,0,0,...,7,1050,910,1965,0,98136,47.5208,-122.393,1360,5000
4,1954400510,20150218T000000,510000,3,2.00,1680,8080,1.0,0,0,...,8,1680,0,1987,0,98074,47.6168,-122.045,1800,7503
5,7237550310,20140512T000000,1225000,4,4.50,5420,101930,1.0,0,0,...,11,3890,1530,2001,0,98053,47.6561,-122.005,4760,101930
6,1321400060,20140627T000000,257500,3,2.25,1715,6819,2.0,0,0,...,7,1715,0,1995,0,98003,47.3097,-122.327,2238,6819
7,2008000270,20150115T000000,291850,3,1.50,1060,9711,1.0,0,0,...,7,1060,0,1963,0,98198,47.4095,-122.315,1650,9711
8,2414600126,20150415T000000,229500,3,1.00,1780,7470,1.0,0,0,...,7,1050,730,1960,0,98146,47.5123,-122.337,1780,8113
9,3793500160,20150312T000000,323000,3,2.50,1890,6560,2.0,0,0,...,7,1890,0,2003,0,98038,47.3684,-122.031,2390,7570


In [4]:
house_ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  int64  
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [ ]:
house_ds.describe()

,id,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
count,2.161300e+04,2.161300e+04,21613.000000,21613.000000,21613.000000,2.161300e+04,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000
mean,4.580302e+09,5.400881e+05,3.370842,2.114757,2079.899736,1.510697e+04,1.494309,0.007542,0.234303,3.409430,7.656873,1788.390691,291.509045,1971.005136,84.402258,98077.939805,47.560053,-122.213896,1986.552492,12768.455652
std,2.876566e+09,3.671272e+05,0.930062,0.770163,918.440897,4.142051e+04,0.539989,0.086517,0.766318,0.650743,1.175459,828.090978,442.575043,29.373411,401.679240,53.505026,0.138564,0.140828,685.391304,27304.179631
min,1.000102e+06,7.500000e+04,0.000000,0.000000,290.000000,5.200000e+02,1.000000,0.000000,0.000000,1.000000,1.000000,290.000000,0.000000,1900.000000,0.000000,98001.000000,47.155900,-122.519000,399.000000,651.000000
25%,2.123049e+09,3.219500e+05,3.000000,1.750000,1427.000000,5.040000e+03,1.000000,0.000000,0.000000,3.000000,7.000000,1190.000000,0.000000,1951.000000,0.000000,98033.000000,47.471000,-122.328000,1490.000000,5100.000000
50%,3.904930e+09,4.500000e+05,3.000000,2.250000,1910.000000,7.618000e+03,1.500000,0.000000,0.000000,3.000000,7.000000,1560.000000,0.000000,1975.000000,0.000000,98065.000000,47.571800,-122.230000,1840.000000,7620.000000
75%,7.308900e+09,6.450000e+05,4.000000,2.500000,2550.000000,1.068800e+04,2.000000,0.000000,0.000000,4.000000,8.000000,2210.000000,560.000000,1997.000000,0.000000,98118.000000,47.678000,-122.125000,2360.000000,10083.000000
max,9.900000e+09,7.700000e+06,33.000000,8.000000,13540.000000,1.651359e+06,3.500000,1.000000,4.000000,5.000000,13.000000,9410.000000,4820.000000,2015.000000,2015.000000,98199.000000,47.777600,-121.315000,6210.000000,871200.000000


# **Q1 :How many houses have a waterfront?**

In [ ]:
# house_ds.where['waterfront'!=0].value_counts()
house_ds.where(house_ds['waterfront']!=0).value_counts()
# house_ds[house_ds.where('waterfront'!=0)].value_counts()

# Ans 163

id            date             price      bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  zipcode  lat      long      sqft_living15  sqft_lot15
1.210290e+08  20140624T000000  549000.0   2.0       1.00       2034.0       13392.0   1.0     1.0         4.0   5.0        7.0    1159.0      875.0          1947.0    0.0           98070.0  47.3312  -122.503  1156.0         15961.0       1
7.403200e+09  20141113T000000  1600000.0  3.0       2.25       3370.0       23065.0   1.0     1.0         4.0   3.0        10.0   1920.0      1450.0         1980.0    0.0           98028.0  47.7434  -122.263  3410.0         19688.0       1
6.329000e+09  20150310T000000  641500.0   1.0       1.00       1000.0       9084.0    1.0     1.0         3.0   3.0        7.0    1000.0      0.0            1950.0    0.0           98146.0  47.5007  -122.382  1090.0         6536.0        1
              20150329T000000  540000.0   3.0

In [ ]:
# loc, iloc and where filters
#loc - used to filter based on the label
#iloc - used to filter based on the position
#where - used to filter based on the condition
house_ds.loc[house_ds['waterfront']!=0]

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
49,822039084,20150311T000000,1350000,3,2.50,2753,65005,1.0,1,2,...,9,2165,588,1953,0,98070,47.4041,-122.451,2680,72513
230,8096000060,20150413T000000,655000,2,1.75,1450,15798,2.0,1,4,...,7,1230,220,1915,1978,98166,47.4497,-122.375,2030,13193
246,2025069065,20140929T000000,2400000,4,2.50,3650,8354,1.0,1,4,...,9,1830,1820,2000,0,98074,47.6338,-122.072,3120,18841
264,2123039032,20141027T000000,369900,1,0.75,760,10079,1.0,1,4,...,5,760,0,1936,0,98070,47.4683,-122.438,1230,14267
300,3225069065,20140624T000000,3075000,4,5.00,4550,18641,1.0,1,4,...,10,2600,1950,2002,0,98074,47.6053,-122.077,4550,19508
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19984,2025069140,20150317T000000,1898000,3,2.50,2830,4334,3.0,1,4,...,10,2830,0,2006,0,98074,47.6318,-122.071,2830,38211
20325,518500480,20140811T000000,3000000,3,3.50,4410,10756,2.0,1,4,...,11,3430,980,2014,0,98056,47.5283,-122.205,3550,5634
20767,8043700105,20150417T000000,2300000,4,4.00,4360,8175,2.5,1,4,...,10,3940,420,2007,0,98008,47.5724,-122.104,2670,8525
21201,518500460,20141008T000000,2230000,3,3.50,3760,5634,2.0,1,4,...,11,2830,930,2014,0,98056,47.5285,-122.205,3560,5762


# **Q2 : Which zip code has the costliest house?**

In [ ]:
# ============ Approach 1 ==============
filter_1 = house_ds['waterfront']!=0
filter_2 = house_ds['price'].max()
# house_ds.loc[house_ds['price'] > 7000000]
house_ds.loc[house_ds['price'] == filter_2]
# house_ds.iloc[]

# Ans 98102

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
7252,6762700020,20141013T000000,7700000,6,8.0,12050,27600,2.5,0,3,...,13,8570,3480,1910,1987,98102,47.6298,-122.323,3940,8800


In [7]:
# ============ Approach 2 ==============
house_ds.head()
house_ds[house_ds['price']==house_ds['price'].max()]['zipcode']

,zipcode
7252,98102


# **Q3 : How many houses are having grade 10?**

In [ ]:
house_ds.loc[house_ds['grade'] == 10].value_counts()
# Ans 1134

id          date             price    bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  zipcode  lat      long      sqft_living15  sqft_lot15
11500240    20150428T000000  872750   3         2.50       2870         13695     2.0     0           0     3          10     2870        0              1991      0             98052    47.6944  -122.102  2840           8472          1
5701700640  20140729T000000  849000   3         3.00       2960         42159     2.0     0           0     3          10     2960        0              1995      0             98052    47.7183  -122.100  2640           25209         1
6072800205  20141119T000000  2375000  4         2.50       3220         20251     1.0     0           0     3          10     3220        0              1969      0             98006    47.5692  -122.192  4200           22114         1
6072800170  20150428T000000  2500000  4         4.00       33

In [12]:
# ============ Approach 1 ==============
house_ds[house_ds['grade'] == 10].count()[0]

<ipython-input-12-edb2519b4c0f>:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  house_ds[house_ds['grade'] == 10].count()[0]


1134

# **Q4 : How many null values are there in the dataset?**

In [ ]:
house_ds.isnull().sum()
# ans 0

id               0
date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

In [13]:
house_ds.isna().sum()
# ans 0

,0
id,0
date,0
price,0
bedrooms,0
bathrooms,0
sqft_living,0
sqft_lot,0
floors,0
waterfront,0
view,0


# **Q5 : Does this 9126100861 customer have a waterfront** ?

In [ ]:
house_ds.tail(5)
filter_1 = house_ds['id']==9126100861
house_ds.loc[filter_1]
# Ans No

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
400,9126100861,20150306T000000,557000,3,3.5,1710,2096,2.0,0,0,...,8,1290,420,2008,0,98122,47.6055,-122.305,1630,1543


In [15]:
# ===== APPROACH 2 =====

if house_ds[house_ds['id']==9126100861] ['waterfront'].values[0] == 0:
  print("No")
else:
  print("Yes")

No


# **Q6 : How many houses have 3 views?**

In [ ]:
filter_1 = house_ds['view']==3
house_ds.loc[filter_1].value_counts()
# Ans 510

id          date             price    bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  zipcode  lat      long      sqft_living15  sqft_lot15
46100204    20150221T000000  1505000  5         3.00       3300         33474     1.0     0           3     3          9      1870        1430           1957      1991          98040    47.5673  -122.210  3836           20953         1
5553300375  20140820T000000  2160000  3         3.50       3080         6495      2.0     0           3     3          11     2530        550            1996      2006          98199    47.6321  -122.393  4120           8620          1
6117500320  20140708T000000  1131000  3         2.25       2790         13791     1.0     0           3     3          8      2790        0              2006      0             98166    47.4389  -122.351  2720           12600         1
6117500160  20150317T000000  425000   3         1.50       15

In [17]:
# ============ Approach 2 ==============
house_ds[house_ds['view']==3].count()[0]

<ipython-input-17-aff4f8bae2c3>:2: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  house_ds[house_ds['view']==3].count()[0]


510

# **Q7 : What is the lowest price of the house?**

In [ ]:
filter_1 = house_ds['price'].min()
house_ds.loc[house_ds['price'] == filter_1]
# Ans 75000

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
1149,3421079032,20150217T000000,75000,1,0.0,670,43377,1.0,0,0,...,3,670,0,1966,0,98022,47.2638,-121.906,1160,42882


In [20]:
# ============ Approach 2 ==============
house_ds.sort_values('price', ascending=True).head(1)['price']

,price
1149,75000


# **Q8: Which zip code area has the cheapest house?**

In [ ]:
filter_1 = house_ds['price'].min()
house_ds.loc[house_ds['price'] == filter_1]
# Ans 98022

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
1149,3421079032,20150217T000000,75000,1,0.0,670,43377,1.0,0,0,...,3,670,0,1966,0,98022,47.2638,-121.906,1160,42882


# **Q9. Which area has the biggest sqft_living?**

In [ ]:
filter_1 = house_ds['sqft_living'].max()
filter_2 = house_ds.loc[house_ds['sqft_living'] == filter_1]
filter_2['zipcode']
# Ans 98053

12777    98053
Name: zipcode, dtype: int64

In [22]:
# ============ Approach 2 ==============
house_ds.sort_values('sqft_living', ascending=False).head(1)['zipcode']

,zipcode
12777,98053


# **Q10 : In which year the costliest house has been built?**

In [ ]:
filter_1 = house_ds['price'].max()
house_ds.loc[house_ds['price'] == filter_1]
# Ans 1910

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
7252,6762700020,20141013T000000,7700000,6,8.0,12050,27600,2.5,0,3,...,13,8570,3480,1910,1987,98102,47.6298,-122.323,3940,8800


# **Q:11. Which of the following data types can a Pandas Series have?**

In [ ]:
series_1 = pd.Series([1,2,3,4,5])
print(series_1)
print(series_1.dtype)
series_2 = pd.Series([1.0,2.0,3,4,5])
print(series_2)
print(series_2.dtype)
series_3 = pd.Series(['Aman','Picku',"Bannu",4,5])
print(series_3)
print(series_3.dtype)

# Ans All Of the Above

0    1
1    2
2    3
3    4
4    5
dtype: int64
int64
0    1.0
1    2.0
2    3.0
3    4.0
4    5.0
dtype: float64
float64
0     Aman
1    Picku
2    Bannu
3        4
4        5
dtype: object
object


# **Q12. Which of the following is having a one-dimensional array?**

In [ ]:
series_1 = pd.Series([[[1,2,3,4,5], [2,4,6,8,10]]])
print(series_1)
print(series_1.dtype)
print(f"dimension:{series_1.ndim}")


df_1 = pd.DataFrame([[[1,2,3,4,5], [2,4,6,8,10]]])
print(df_1)
print(f"dimension:{df_1.ndim}")
# Ans Series

0    [[1, 2, 3, 4, 5], [2, 4, 6, 8, 10]]
dtype: object
object
dimension:1
                 0                 1
0  [1, 2, 3, 4, 5]  [2, 4, 6, 8, 10]
dimension:2


# **Q13. What value will you get on the left side after printing a series format data?**

In [ ]:
# Ans Index

# **Q13:Which of the following is considered as data in pandas?**

In [ ]:
# Ans All of the Above

# **Q15 : In what format the keys will get converted into when we convert the dictionary data into**

In [ ]:
dict_data = {1:['a', 'b','c'],
                2:[1,2,3],
             3:[1.0, 2.0, 3.0]}
print(dict_data)
print(type(dict_data))

#  changing the dict to dataframe
df_1 = pd.DataFrame(dict_data)
print(df_1)
print(type(df_1))

series_1 = pd.Series([[[1,2,3,4,5], [2,4,6,8,10]]])
print(series_1)
# changing the series to dataframe
df_2 = pd.DataFrame(series_1)
print(df_2)
print(type(df_2))

# Ans Column


{1: ['a', 'b', 'c'], 2: [1, 2, 3], 3: [1.0, 2.0, 3.0]}
<class 'dict'>
   1  2    3
0  a  1  1.0
1  b  2  2.0
2  c  3  3.0
<class 'pandas.core.frame.DataFrame'>
0    [[1, 2, 3, 4, 5], [2, 4, 6, 8, 10]]
dtype: object
                                     0
0  [[1, 2, 3, 4, 5], [2, 4, 6, 8, 10]]
<class 'pandas.core.frame.DataFrame'>


# **Q16 : Write a function to perform the following operation:**
Increase the grade of the house by 1 if the sqft of the house is greater than 0 and less than
equal to 400. If the total sqft of the house is greater than 400, increase the grade of the house
by 2.
After creating the above function, which syntax will be using when using apply function to run
the above created function in the dataset?

In [ ]:
def grade_increase(x):
    if x['sqft_living'] > 0 and x['sqft_living'] <= 400:
        x['grade'] += 1
    elif x['sqft_living'] > 400:
        x['grade'] += 2

    return x

In [ ]:
print(house_ds.head(10))
house_ds_1 = house_ds.apply(grade_increase, axis=1)
print(house_ds_1.head(10))
# house_ds_1 = grade_increase(house_ds)
# Ans data.apply(your_function_name(arg1))

           id             date    price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900         3       1.00         1180   
1  6414100192  20141209T000000   538000         3       2.25         2570   
2  5631500400  20150225T000000   180000         2       1.00          770   
3  2487200875  20141209T000000   604000         4       3.00         1960   
4  1954400510  20150218T000000   510000         3       2.00         1680   
5  7237550310  20140512T000000  1225000         4       4.50         5420   
6  1321400060  20140627T000000   257500         3       2.25         1715   
7  2008000270  20150115T000000   291850         3       1.50         1060   
8  2414600126  20150415T000000   229500         3       1.00         1780   
9  3793500160  20150312T000000   323000         3       2.50         1890   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180           

**IF WE WANT THE FUNCTION TO WORK ON EVERY ROW THEN WE MUST USE APPLY, OTHERWISE WE CAN DIRECTLY CALL THE FUNCTION BY NAME**

In [42]:
# ============ Approach 2 ==============
def price_inscrease(x):
    if x['sqft_living'] > 0 and x['sqft_living'] <= 400:
        x['grade'] += 1
        return x['grade']
    elif x['sqft_living'] > 400:
        x['grade'] += 2
        return x['grade']

house_ds['grade_new'] = house_ds.apply(price_inscrease, axis=1)

# price_inscrease(house_ds)
house_ds[['grade', 'grade_new', 'sqft_living']].head(20)


,grade,grade_new,sqft_living
0,7,9,1180
1,7,9,2570
2,6,8,770
3,7,9,1960
4,8,10,1680
5,11,13,5420
6,7,9,1715
7,7,9,1060
8,7,9,1780
9,7,9,1890


# **Q17: How to check the duplicate values on id,grade and location?**

In [ ]:
# house_ds_nd = house_ds.duplicated(subset=['id','grade','location'])
# house_ds.tail(10)
# house_ds.tail(10)
house_ds[house_ds.duplicated(subset=['id','grade'])]
# house_ds[house_ds.duplicated(subset=['grade'])].value_counts()
# house_ds.duplicated(subset=['grade']).value_counts()
# house_ds.duplicate(subset=['grade'])

# Ans A


,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,Floor Area
94,6021501535,20141223T000000,700000,3,1.50,1580,5000,1.0,0,0,...,1290,290,1939,0,98117,47.6870,-122.386,1570,4500,8160
314,4139480200,20141209T000000,1400000,4,3.25,4290,12103,1.0,0,3,...,2690,1600,1997,0,98006,47.5503,-122.102,3860,11244,20683
325,7520000520,20150311T000000,240500,2,1.00,1240,12092,1.0,0,0,...,960,280,1922,1984,98146,47.4957,-122.352,1820,7460,14572
346,3969300030,20141229T000000,239900,4,1.00,1000,7134,1.0,0,0,...,1000,0,1943,0,98178,47.4897,-122.240,1020,7138,9134
372,2231500030,20150324T000000,530000,4,2.25,2180,10754,1.0,0,0,...,1100,1080,1954,0,98133,47.7711,-122.341,1810,6929,15114
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
20181,7853400250,20150219T000000,645000,4,3.50,2910,5260,2.0,0,0,...,2910,0,2012,0,98065,47.5168,-121.883,2910,5260,11080
20613,2724049222,20141201T000000,220000,2,2.50,1000,1092,2.0,0,0,...,990,10,2004,0,98118,47.5419,-122.271,1330,1466,3092
20670,8564860270,20150330T000000,502000,4,2.50,2680,5539,2.0,0,0,...,2680,0,2013,0,98045,47.4759,-121.734,2680,5992,10899
20780,6300000226,20150504T000000,380000,4,1.00,1200,2171,1.5,0,0,...,1200,0,1933,0,98133,47.7076,-122.342,1130,1598,4571


In [54]:
# ============== APPROACH 2 ====================
# house_ds.duplicated(subset=['id', 'grade', 'zipcode'])
# house_ds[house_ds.duplicated(subset=['id', 'grade', 'zipcode'])]
house_ds[house_ds.duplicated(['id', 'grade', 'zipcode'])]
# house_ds.duplicate(subset=['id','zipcode','grade'])

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,grade_new
94,6021501535,20141223T000000,700000,3,1.50,1580,5000,1.0,0,0,...,1290,290,1939,0,98117,47.6870,-122.386,1570,4500,10
314,4139480200,20141209T000000,1400000,4,3.25,4290,12103,1.0,0,3,...,2690,1600,1997,0,98006,47.5503,-122.102,3860,11244,13
325,7520000520,20150311T000000,240500,2,1.00,1240,12092,1.0,0,0,...,960,280,1922,1984,98146,47.4957,-122.352,1820,7460,8
346,3969300030,20141229T000000,239900,4,1.00,1000,7134,1.0,0,0,...,1000,0,1943,0,98178,47.4897,-122.240,1020,7138,8
372,2231500030,20150324T000000,530000,4,2.25,2180,10754,1.0,0,0,...,1100,1080,1954,0,98133,47.7711,-122.341,1810,6929,9
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
20181,7853400250,20150219T000000,645000,4,3.50,2910,5260,2.0,0,0,...,2910,0,2012,0,98065,47.5168,-121.883,2910,5260,11
20613,2724049222,20141201T000000,220000,2,2.50,1000,1092,2.0,0,0,...,990,10,2004,0,98118,47.5419,-122.271,1330,1466,9
20670,8564860270,20150330T000000,502000,4,2.50,2680,5539,2.0,0,0,...,2680,0,2013,0,98045,47.4759,-121.734,2680,5992,10
20780,6300000226,20150504T000000,380000,4,1.00,1200,2171,1.5,0,0,...,1200,0,1933,0,98133,47.7076,-122.342,1130,1598,9


# **Q18. Which of the following code will help to display the 3rd, 4th and 5th rows from the 6th to 9th**
columns of data frame data?

In [ ]:
house_ds.iloc[3:6, 6:10]

,sqft_lot,floors,waterfront,view
3,5000,1.0,0,0
4,8080,1.0,0,0
5,101930,1.0,0,0


# Q19. Which of the following syntax will display the last two records of df?
# import pandas as pd **bold text**

In [ ]:
# house_ds.tail(2)
house_ds.iloc[-2:]

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
21611,291310100,20150116T000000,400000,3,2.50,1600,2388,2.0,0,0,...,8,1600,0,2004,0,98027,47.5345,-122.069,1410,1287
21612,1523300157,20141015T000000,325000,2,0.75,1020,1076,2.0,0,0,...,7,1020,0,2008,0,98144,47.5941,-122.299,1020,1357


In [ ]:
df = pd.DataFrame({'A':[34, 78, 54], 'B':[12, 67, 43]}, index=['r1', 'r2', 'r3'])
df
df.loc['r2':'r3']

# ans df.loc['r2':'r3']


,A,B
r2,78,67
r3,54,43


In [55]:
# ====== Approach 2 ====
df = pd.DataFrame({'A':[34, 78, 54], 'B':[12, 67, 43]}, index=['r1', 'r2', 'r3'])

In [61]:
df.loc['r2': 'r3']

,A,B
r2,78,67
r3,54,43


# **If the index is in String you must use Loc instead of iLoc for retriving the details**

# **Q20 Which of the following is/are true about loc in pandas:**

In [82]:
# A. Add new rows in the data frames
# B. To change the values of a row to a particular value
# C. To extract values from the particular rows
# D. All of the above

df = pd.DataFrame({'A':[34, 78, 54], 'B':[12, 67, 43]}, index=['r1', 'r2', 'r3'])
new_df = pd.DataFrame({'A':[34, 78, 54], 'B':[12, 67, 43], 'C': [23, 54,56]}, index=['r1', 'r2', 'r3'])
new_df

new_df.loc['r1'] = 0
new_df


new_df.loc['r5'] = [23, 54,56]
new_df

,A,B,C
r1,0,0,0
r2,78,67,54
r3,54,43,56
r5,23,54,56


# **Q21 : Change the date column in the format (dd/mm/year) using the pandas to_datetime()**

In [ ]:
# Ans pd.to_datetime(data['date'], format='%Y-%m-%d', utc=False, dayfirst=True)

# **Q22. Create a separate data frame that satisfies the conditions below.**
1. Houses built before 1980
2. Have more than 2 bedrooms
3. Have more than 2 floors

In [ ]:
house_ds.loc[(house_ds["yr_built"] < 1980) & (house_ds['floors'] > 2) & (house_ds['bedrooms'] > 2)]

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
518,4386700135,20141114T000000,2250000,4,2.25,4760,8036,2.5,0,0,...,9,3390,1370,1916,0,98112,47.6415,-122.285,2950,9323
759,770000045,20141024T000000,405600,5,1.50,2830,4000,2.5,0,0,...,8,2830,0,1918,0,98118,47.5132,-122.262,1480,4000
960,7578200310,20141112T000000,650000,4,2.00,2208,5000,3.0,0,0,...,8,2208,0,1917,0,98116,47.5711,-122.383,1760,5000
1296,5458800125,20140514T000000,925000,4,2.50,2190,7350,2.5,0,0,...,8,2190,0,1958,0,98040,47.5786,-122.236,1880,7350
1350,7234600786,20140511T000000,842500,4,2.50,2160,5298,2.5,0,0,...,9,2160,0,1902,0,98122,47.6106,-122.310,1720,2283
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
18535,3667500015,20140925T000000,770000,4,3.50,3680,2242,2.5,0,0,...,9,2670,1010,1930,2007,98112,47.6192,-122.307,1350,1288
18631,1337800665,20140811T000000,1325000,4,3.25,2850,4800,2.5,0,0,...,10,2700,150,1905,0,98112,47.6292,-122.312,2850,4800
19068,6613001241,20140811T000000,1415000,4,3.00,3110,4408,2.5,0,3,...,10,2510,600,1931,0,98105,47.6583,-122.270,3250,5669
19188,7942601475,20140520T000000,345600,5,3.50,2800,5120,2.5,0,0,...,9,2800,0,1903,2005,98122,47.6059,-122.310,1780,5120


# **Q23. For a given nested list, convert the same into a dataframe.**
sample_list = [['Carl', 22],
['Martha', 25],
['Calvin', 12],
['Stuart', 15]
#

The resulting dataframe must contain the column names as ‘Name’, and ‘Age’ with the
respective values from the sample_list.

In [ ]:
sample_list = [['Carl', 22], ['Martha', 25], ['Calvin', 12], ['Stuart', 15]]
pd.DataFrame(sample_list, columns=['Name', 'Age'])

,Name,Age
0,Carl,22
1,Martha,25
2,Calvin,12
3,Stuart,15


# **Q24 : For a given dictionary, convert the same into a dataframe.**

In [ ]:
sample_dict = {'Cristiano': ['Ronaldo','Man U', 801],'Lionel': ['Messi','PSG', 758],'Luis': ['Suarez','Atletico Madrid', 509],'Robert': ['Lewandowski','Bayern Munich', 527],'Zlatan': ['Ibrahimovic','AC Milan',553]}

df_1 = pd.DataFrame(sample_dict)
df_1 = df_1.transpose()
df_1.reset_index(inplace = True)
df_1.columns = ['First Name','Last Name','Club','Goals']
print(df_1)

  First Name    Last Name             Club Goals
0  Cristiano      Ronaldo            Man U   801
1     Lionel        Messi              PSG   758
2       Luis       Suarez  Atletico Madrid   509
3     Robert  Lewandowski    Bayern Munich   527
4     Zlatan  Ibrahimovic         AC Milan   553


# **Q25 : For a given tuple, convert the same into a dataframe**

In [ ]:
sample_tuple = ([1, 'one', 3],[2, 'two', 3],[3, 'Three', 5],[4, 'Four', 4],[5, 'Five', 4])
df_1 = pd.DataFrame(sample_tuple, columns=['Number','Word','Length'])
print(df_1)

# ans C

   Number   Word  Length
0       1    one       3
1       2    two       3
2       3  Three       5
3       4   Four       4
4       5   Five       4


In [109]:
# ===== Approach 2 =============
sample_tuple = ([1, 'one', 3], [2, 'two', 3], [3, 'Three', 5], [4, 'Four', 4], [5, 'Five', 4])
df_1 = pd.DataFrame(sample_tuple, columns=['Number', 'Word', 'Length'])
df_1

,Number,Word,Length
0,1,one,3
1,2,two,3
2,3,Three,5
3,4,Four,4
4,5,Five,4


# **Q26: Create a separate dataframe that contains houses ordered in ascending or descending**
**# order of the prices of each house.**

In [ ]:
ascending = house_ds.sort_values('price', ascending=False)
descending = house_ds.sort_values('price', ascending=True)
print(ascending)
print(descending)

               id             date    price  bedrooms  bathrooms  sqft_living  \
7252   6762700020  20141013T000000  7700000         6       8.00        12050   
3914   9808700762  20140611T000000  7062500         5       4.50        10040   
9254   9208900037  20140919T000000  6885000         6       7.75         9890   
4411   2470100110  20140804T000000  5570000         5       5.75         9200   
1448   8907500070  20150413T000000  5350000         5       5.00         8000   
...           ...              ...      ...       ...        ...          ...   
8274   3883800011  20141105T000000    82000         3       1.00          860   
16198  3028200080  20150324T000000    81000         2       1.00          730   
465    8658300340  20140523T000000    80000         1       0.75          430   
15293    40000362  20140506T000000    78000         2       1.00          780   
1149   3421079032  20150217T000000    75000         1       0.00          670   

       sqft_lot  floors  wa

In [111]:
# ==== Approach 2 ==============
hprice_ascend = house_ds.sort_values('price', ascending=True)
hprice_ascend

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,grade_new
1149,3421079032,20150217T000000,75000,1,0.00,670,43377,1.0,0,0,...,670,0,1966,0,98022,47.2638,-121.906,1160,42882,5
15293,40000362,20140506T000000,78000,2,1.00,780,16344,1.0,0,0,...,780,0,1942,0,98168,47.4739,-122.280,1700,10387,7
465,8658300340,20140523T000000,80000,1,0.75,430,5050,1.0,0,0,...,430,0,1912,0,98014,47.6499,-121.909,1200,7500,6
16198,3028200080,20150324T000000,81000,2,1.00,730,9975,1.0,0,0,...,730,0,1943,0,98168,47.4808,-122.315,860,9000,7
8274,3883800011,20141105T000000,82000,3,1.00,860,10426,1.0,0,0,...,860,0,1954,0,98146,47.4987,-122.341,1140,11250,8
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1448,8907500070,20150413T000000,5350000,5,5.00,8000,23985,2.0,0,4,...,6720,1280,2009,0,98004,47.6232,-122.220,4600,21750,14
4411,2470100110,20140804T000000,5570000,5,5.75,9200,35069,2.0,0,0,...,6200,3000,2001,0,98039,47.6289,-122.233,3560,24345,15
9254,9208900037,20140919T000000,6885000,6,7.75,9890,31374,2.0,0,4,...,8860,1030,2001,0,98039,47.6305,-122.240,4540,42730,15
3914,9808700762,20140611T000000,7062500,5,4.50,10040,37325,2.0,1,2,...,7680,2360,1940,2001,98004,47.6500,-122.214,3930,25449,13


# **Q27. Calculate the mean and standard deviation of all the numerical values in the dataset. For**
example - the mean for the bedrooms column is 3.370 and the standard deviation is 0.930.

In [ ]:
house_ds.describe()

,id,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
count,2.161300e+04,2.161300e+04,21613.000000,21613.000000,21613.000000,2.161300e+04,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000,21613.000000
mean,4.580302e+09,5.400881e+05,3.370842,2.114757,2079.899736,1.510697e+04,1.494309,0.007542,0.234303,3.409430,7.656873,1788.390691,291.509045,1971.005136,84.402258,98077.939805,47.560053,-122.213896,1986.552492,12768.455652
std,2.876566e+09,3.671272e+05,0.930062,0.770163,918.440897,4.142051e+04,0.539989,0.086517,0.766318,0.650743,1.175459,828.090978,442.575043,29.373411,401.679240,53.505026,0.138564,0.140828,685.391304,27304.179631
min,1.000102e+06,7.500000e+04,0.000000,0.000000,290.000000,5.200000e+02,1.000000,0.000000,0.000000,1.000000,1.000000,290.000000,0.000000,1900.000000,0.000000,98001.000000,47.155900,-122.519000,399.000000,651.000000
25%,2.123049e+09,3.219500e+05,3.000000,1.750000,1427.000000,5.040000e+03,1.000000,0.000000,0.000000,3.000000,7.000000,1190.000000,0.000000,1951.000000,0.000000,98033.000000,47.471000,-122.328000,1490.000000,5100.000000
50%,3.904930e+09,4.500000e+05,3.000000,2.250000,1910.000000,7.618000e+03,1.500000,0.000000,0.000000,3.000000,7.000000,1560.000000,0.000000,1975.000000,0.000000,98065.000000,47.571800,-122.230000,1840.000000,7620.000000
75%,7.308900e+09,6.450000e+05,4.000000,2.500000,2550.000000,1.068800e+04,2.000000,0.000000,0.000000,4.000000,8.000000,2210.000000,560.000000,1997.000000,0.000000,98118.000000,47.678000,-122.125000,2360.000000,10083.000000
max,9.900000e+09,7.700000e+06,33.000000,8.000000,13540.000000,1.651359e+06,3.500000,1.000000,4.000000,5.000000,13.000000,9410.000000,4820.000000,2015.000000,2015.000000,98199.000000,47.777600,-121.315000,6210.000000,871200.000000


In [ ]:
house_ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  int64  
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [ ]:
house_ds.std()

TypeError: could not convert string to float: '20141013T000000'

# **Q28 : 28. Perform the following operations on the pandas dataframe.**
1. Get the data starting from the rows 25 to 35.
2. Get the price of the houses located at the longitude--122.045 latitude- 47.6168.

In [ ]:
house_ds.iloc[25:36]

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,grade,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15
25,1202000200,20141103T000000,233000,3,2.00,1710,4697,1.5,0,0,...,6,1710,0,1941,0,98002,47.3048,-122.218,1030,4705
26,1794500383,20140626T000000,937000,3,1.75,2450,2691,2.0,0,0,...,8,1750,700,1915,0,98119,47.6386,-122.360,1760,3573
27,3303700376,20141201T000000,667000,3,1.00,1400,1581,1.5,0,0,...,8,1400,0,1909,0,98112,47.6221,-122.314,1860,3861
28,5101402488,20140624T000000,438000,3,1.75,1520,6380,1.0,0,0,...,7,790,730,1948,0,98115,47.6950,-122.304,1520,6235
29,1873100390,20150302T000000,719000,4,2.50,2570,7173,2.0,0,0,...,8,2570,0,2005,0,98052,47.7073,-122.110,2630,6026
30,8562750320,20141110T000000,580500,3,2.50,2320,3980,2.0,0,0,...,8,2320,0,2003,0,98027,47.5391,-122.070,2580,3980
31,2426039314,20141201T000000,280000,2,1.50,1190,1265,3.0,0,0,...,7,1190,0,2005,0,98133,47.7274,-122.357,1390,1756
32,461000390,20140624T000000,687500,4,1.75,2330,5000,1.5,0,0,...,7,1510,820,1929,0,98117,47.6823,-122.368,1460,5000
33,7589200193,20141110T000000,535000,3,1.00,1090,3000,1.5,0,0,...,8,1090,0,1929,0,98117,47.6889,-122.375,1570,5080
34,7955080270,20141203T000000,322500,4,2.75,2060,6659,1.0,0,0,...,7,1280,780,1981,0,98058,47.4276,-122.157,2020,8720


In [ ]:
filter_1 = house_ds.loc[(house_ds["lat"] == 47.6168) & (house_ds["long"] == -122.045)]
filter_1["price"]

4    510000
Name: price, dtype: int64

# **Q29 : Create a new column with the floor area(sqft_living, sqft_lot, sqft_above, sqft_basement, all**
combined in one column).

In [ ]:
house_ds['Floor Area'] = house_ds['sqft_living'] + house_ds['sqft_lot'] + house_ds['sqft_above'] + house_ds['sqft_basement']
house_ds.head(10)

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,Floor Area
0,7129300520,20141013T000000,221900,3,1.00,1180,5650,1.0,0,0,...,1180,0,1955,0,98178,47.5112,-122.257,1340,5650,8010
1,6414100192,20141209T000000,538000,3,2.25,2570,7242,2.0,0,0,...,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639,12382
2,5631500400,20150225T000000,180000,2,1.00,770,10000,1.0,0,0,...,770,0,1933,0,98028,47.7379,-122.233,2720,8062,11540
3,2487200875,20141209T000000,604000,4,3.00,1960,5000,1.0,0,0,...,1050,910,1965,0,98136,47.5208,-122.393,1360,5000,8920
4,1954400510,20150218T000000,510000,3,2.00,1680,8080,1.0,0,0,...,1680,0,1987,0,98074,47.6168,-122.045,1800,7503,11440
5,7237550310,20140512T000000,1225000,4,4.50,5420,101930,1.0,0,0,...,3890,1530,2001,0,98053,47.6561,-122.005,4760,101930,112770
6,1321400060,20140627T000000,257500,3,2.25,1715,6819,2.0,0,0,...,1715,0,1995,0,98003,47.3097,-122.327,2238,6819,10249
7,2008000270,20150115T000000,291850,3,1.50,1060,9711,1.0,0,0,...,1060,0,1963,0,98198,47.4095,-122.315,1650,9711,11831
8,2414600126,20150415T000000,229500,3,1.00,1780,7470,1.0,0,0,...,1050,730,1960,0,98146,47.5123,-122.337,1780,8113,11030
9,3793500160,20150312T000000,323000,3,2.50,1890,6560,2.0,0,0,...,1890,0,2003,0,98038,47.3684,-122.031,2390,7570,10340


In [121]:
# house_ds['Floor Area'] = house_ds['sqft_living'] + house_ds['sqft_lot'] + house_ds['sqft_above'] + house_ds['sqft_basement']
# house_ds.head(10)


house_ds['Floor Area'] = house_ds['sqft_living', 'sqft_lot','sqft_basement','sqft_above']

ValueError: Columns must be same length as key

In [114]:
house_ds['Floor Area'] = house_ds['sqft_living'], house_ds['sqft_lot'], house_ds['sqft_basement'], house_ds['sqft_above']

ValueError: Length of values (4) does not match length of index (21613)

**23) Perform the following operations on the dataframes given below. A = pd.DataFrame([['Carl', 22],['Martha', 25],['Calvin', 12],['Stuart', 15]], columns=[‘Name’, ‘Age’]) B = pd.DataFrame([['Melvin', 25],['Martha', 34],['Lewis', 32],['Leo', 25]], columns=[‘Name’, ‘Age’]) 1. Left Outer Join 2. Outer Join 3. Inner Join 4. Right Outer Join**

In [126]:
A = pd.DataFrame([['Carl', 22],['Martha', 25],['Calvin', 12],['Stuart', 15]], columns=['Name','Age'])
B = pd.DataFrame([['Melvin', 25],['Martha', 34],['Lewis', 32],['Leo', 25]], columns=['Name', 'Age'])

pd.merge(A, B, on='Age', how='left')
pd.merge(A, B, on='Age', how='outer')
pd.merge(A, B, on='Age', how='inner')
pd.merge(A, B, on='Age', how='right')

,Name_x,Age,Name_y
0,Martha,25,Melvin
1,NaN,34,Martha
2,NaN,32,Lewis
3,Martha,25,Leo


In [128]:
house_ds.corr(numeric_only=True)

,id,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,condition,...,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,grade_new,Floor Area
id,1.000000,-0.016762,0.001286,0.005160,-0.012258,-0.132109,0.018525,-0.002721,0.011592,-0.023783,...,-0.005151,0.021380,-0.016907,-0.008224,-0.001891,0.020799,-0.002901,-0.138798,0.008145,-0.131520
price,-0.016762,1.000000,0.308350,0.525138,0.702035,0.089661,0.256794,0.266369,0.397293,0.036362,...,0.323816,0.054012,0.126434,-0.053203,0.307003,0.021626,0.585379,0.082447,0.667097,0.119763
bedrooms,0.001286,0.308350,1.000000,0.515884,0.576671,0.031703,0.175429,-0.006582,0.079532,0.028472,...,0.303093,0.154178,0.018841,-0.152668,-0.008931,0.129473,0.391638,0.029244,0.357409,0.056788
bathrooms,0.005160,0.525138,0.515884,1.000000,0.754665,0.087740,0.500653,0.063744,0.187737,-0.124982,...,0.283770,0.506019,0.050739,-0.203866,0.024573,0.223042,0.568634,0.087175,0.664881,0.120172
sqft_living,-0.012258,0.702035,0.576671,0.754665,1.000000,0.172826,0.353949,0.103818,0.284611,-0.058753,...,0.435043,0.318049,0.055363,-0.199430,0.052529,0.240223,0.756420,0.183286,0.762534,0.215319
sqft_lot,-0.132109,0.089661,0.031703,0.087740,0.172826,1.000000,-0.005201,0.021604,0.074710,-0.008958,...,0.015286,0.053080,0.007644,-0.129574,-0.085683,0.229521,0.144608,0.718557,0.113370,0.999062
floors,0.018525,0.256794,0.175429,0.500653,0.353949,-0.005201,1.000000,0.023698,0.029444,-0.263768,...,-0.245705,0.489319,0.006338,-0.059121,0.049614,0.125419,0.279885,-0.011269,0.458032,0.010406
waterfront,-0.002721,0.266369,-0.006582,0.063744,0.103818,0.021604,0.023698,1.000000,0.401857,0.016653,...,0.080588,-0.026161,0.092885,0.030285,-0.014274,-0.041910,0.086463,0.030703,0.082729,0.025984
view,0.011592,0.397293,0.079532,0.187737,0.284611,0.074710,0.029444,0.401857,1.000000,0.045990,...,0.276947,-0.053440,0.103917,0.084827,0.006157,-0.078400,0.280439,0.072575,0.251191,0.086586
condition,-0.023783,0.036362,0.028472,-0.124982,-0.058753,-0.008958,-0.263768,0.016653,0.045990,1.000000,...,0.174105,-0.361417,-0.060618,0.003026,-0.014941,-0.106500,-0.092824,-0.003406,-0.144409,-0.011465


30. Perform the following operations on the dataframes given below.
A = pd.DataFrame([['Carl', 22],['Martha', 25],['Calvin', 12],['Stuart', 15]], columns=[‘Name’, ‘Age’])
# **B = pd.DataFrame([['Melvin', 25],['Martha', 34],['Lewis', 32],['Leo', 25]], columns=[‘Name’, ‘Age’])**

In [ ]:
A = pd.DataFrame([['Carl', 22],['Martha', 25],['Calvin', 12],['Stuart', 15]], columns=['Name', 'Age'])
B = pd.DataFrame([['Melvin', 25],['Martha', 34],['Lewis', 32],['Leo', 25]], columns=['Name', 'Age'])
print(A)
print(B)

     Name  Age
0    Carl   22
1  Martha   25
2  Calvin   12
3  Stuart   15
     Name  Age
0  Melvin   25
1  Martha   34
2   Lewis   32
3     Leo   25


In [ ]:
inner = pd.merge(A, B, on='Age', how='inner')
outer = pd.merge(A, B, on='Age', how='outer')
left_outer = pd.merge(A, B, on='Age', how='left')
right_outer = pd.merge(A, B, on='Age', how='right')

# Q31 **What will be the correlation between the columns sqft_living and sqft_above?**


In [ ]:
correlation = house_ds['sqft_living'].corr(house_ds['sqft_above'])
print(correlation)
# Ans 08765

0.876596598681318


# **Q32. If the correlation between the columns bathrooms and sqft_living is 0.754665, what all interpretations can be made about the two columns?**
**A. A positive correlation between the two columns**
B. The columns show perfect correlation.
C. A strong negative correlation between the two columns
D. No correlation between the two columns.

Class Questions :  Create pandas series

In [ ]:
import pandas as pd
import numpy as np

list_1 = [1,2,3,4,5]
tuple_1 = (1,2,3,4,5)
set_1 = {1,2,3,4,5}
dict_1 = {'a':1, 'b':2, 'c':3, 'd':4, 'e':5}

series_list = pd.Series(list_1)
series_tuple = pd.Series(tuple_1)
series_dict = pd.Series(dict_1)
series_set = pd.Series(list(set_1))

print(series_list)
print(series_tuple)
print(series_dict)
print(series_set)

print(series_list)
print(series_tuple)
print(series_tuple)

0    1
1    2
2    3
3    4
4    5
dtype: int64
0    1
1    2
2    3
3    4
4    5
dtype: int64
a    1
b    2
c    3
d    4
e    5
dtype: int64
0    1
1    2
2    3
3    4
4    5
dtype: int64
0    1
1    2
2    3
3    4
4    5
dtype: int64
0    1
1    2
2    3
3    4
4    5
dtype: int64
0    1
1    2
2    3
3    4
4    5
dtype: int64


**Question** Create a python dictionary with three key value pairs. Keep the keys as the first names, and the respective last names as values in the dictionary. Use the dictionary to create a pandas series.

In [ ]:
disct_First_lastName = {'Prasanna': "Kumar", "Prianshu": "Kumar"}
print(disct_First_lastName)
print(type(disct_First_lastName))

series_disct_First_lastName = pd.Series(disct_First_lastName.values(), index=disct_First_lastName.keys())
series_disct_First_lastName

{'Prasanna': 'Kumar', 'Prianshu': 'Kumar'}
<class 'dict'>


,0
Prasanna,Kumar
Prianshu,Kumar


Create a pandas series, with the elements as numbers from 1-10. Use the index values starting from 0-10. For the above series:

- Convert it to a list.
- Convert it to a dictionary.
- Convert it to a tuple.
- Convert it to a set.

In [ ]:
series_1 = pd.Series([1,2,3,4,5,6,7,8,9,10])
list_1 = list(series_1.index)
print(list_1)


dict_1 = series_1.to_dict()
print(dict_1)

tuple_1 = tuple(series_1.index)
print(tuple_1)

set_1 = set(series_1.index)
print(set_1)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
{0: 1, 1: 2, 2: 3, 3: 4, 4: 5, 5: 6, 6: 7, 7: 8, 8: 9, 9: 10}
(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}


**Question** : Create a sample pandas series with x integer elements. Perform the following operations on the same:

- Print the first 5 elements of the series
- Print the last 5 elements of the series
- Insert an integer value ‘50’ at the index 5.
- Update the last element in the series with the value 100.
- Delete the first element from the series.



Note: The index values must not be affected by the operations and should be uniform for the series after the last operation.

In [ ]:
input_num_element = input("enter numbers :")

series_1 = pd.Series(range(1, int(input_num_element)+1))
print(series_1)

print(series_1.head(5))
print(series_1.tail(5))

# series_1.loc['5'] = 50
# print(series_1)
list_1 = list(series_1)


list_1.insert(5, 50)
print(list_1)
series_1 = pd.Series(list_1)
print(series_1)

series_1[len(series_1)-1] = 100
print(series_1)


enter numbers :10
0     1
1     2
2     3
3     4
4     5
5     6
6     7
7     8
8     9
9    10
dtype: int64
0    1
1    2
2    3
3    4
4    5
dtype: int64
5     6
6     7
7     8
8     9
9    10
dtype: int64
[1, 2, 3, 4, 5, 50, 6, 7, 8, 9, 10]
0      1
1      2
2      3
3      4
4      5
5     50
6      6
7      7
8      8
9      9
10    10
dtype: int64
0       1
1       2
2       3
3       4
4       5
5      50
6       6
7       7
8       8
9       9
10    100
dtype: int64


For a given pandas series obtained in the 4th step. Write a sample code to find the element 100 in the series.

In [ ]:
var = 2
(series_1[series_1 == var].index)[0]

print(f" at index {(series_1[series_1 == var].index)[0]} element {var} is present")

 at index 1 element 2 is present


For the given series A and B, perform concatenation on the same.

series_A = pd.Series([1,2,3,4,5])

series_B = pd.Series([6,7,8,9,10])

In [ ]:
series_A = pd.Series([1,2,3,4,5])
series_B = pd.Series([6,7,8,9,10])
series_C = pd.concat([series_A, series_B], ignore_index=True)
print(series_C)

0     1
1     2
2     3
3     4
4     5
5     6
6     7
7     8
8     9
9    10
dtype: int64


**Question**  :
For a given nested list and a sample dictionary, create two data frames using pandas.


List = [['AA','BB','CC'], ['DD','EE','FF'], ['GG','HH','II']]


Dictionary = { 'Kris': ['Jordan', 12,22],'Memphis': ['Depay', 23, 24]}

In [ ]:
li_1 = [['AA','BB','CC'], ['DD','EE','FF'], ['GG','HH','II']]
di_1 = { 'Kris': ['Jordan', 12,22],'Memphis': ['Depay', 23, 24]}

df_1 = pd.DataFrame(li_1)
df_2 = pd.DataFrame(di_1)
print(df_1)
print(df_2)


    0   1   2
0  AA  BB  CC
1  DD  EE  FF
2  GG  HH  II
     Kris Memphis
0  Jordan   Depay
1      12      23
2      22      24


Create a pandas Dataframe with two columns = [‘Name’, ‘Age’]. Perform the following operations on the same.


Print the first 3 rows of the Data Frame .


Print the rows that have the age greater than 15.


Insert 5 imaginary names and ages as values to the pandas Data Frame .


Update the 5th row so that the name = john and age=56.


Delete the first row from the Data Frame .

In [ ]:
d = { 'Name': ['Jordan', 'John','Alex'],'Age': [35, 23, 24]}
df_1 = pd.DataFrame(d)
print(df_1)

print(df_1.head(3))

print(df_1[df_1['Age'] > 15])

     Name  Age
0  Jordan   35
1    John   23
2    Alex   24
     Name  Age
0  Jordan   35
1    John   23
2    Alex   24
     Name  Age
0  Jordan   35
2    Alex   24


In [ ]:
d = { 'Name': ['pk', 'jk','kk'],'Age': [35, 23, 24]}
df_2 = pd.DataFrame(d)
pd.concat([df_1, df_2], ignore_index=True)
# df_3 = df_1.append(df_2, ignore_index=True)
# print(df_3)
df_1.drop('Age', axis=1, inplace=True) #deleting the
print(df_1)

   Name
1  John
2  Alex


Question Data frame to list and Dataframe to dictionary

In [ ]:
list_1 = df_2.values.tolist()
print(list_1)
print(type(list_1))

dict_1 = df_2.to_dict()
print(dict_1)
print(type(dict_1))

[['pk', 35], ['jk', 23], ['kk', 24]]
<class 'list'>
{'Name': {0: 'pk', 1: 'jk', 2: 'kk'}, 'Age': {0: 35, 1: 23, 2: 24}}
<class 'dict'>


Create two separate Data Frame s A and B, and perform the following operations on them.

Concatenate the two Data Frame s to a new Data Frame C.

Split the Data Frame C row wise.

Split the Data Frame column wise based on a condition. For example - columns values greater than 5, column values less than 5, etc.

In [ ]:
df_1 = pd.DataFrame({'A': [1, 2, 3], 'B': [4, 5, 6]})
df_2 = pd.DataFrame({'A': [7, 8, 9], 'B': [10, 11, 12]})
df_3 = pd.concat([df_1, df_2], ignore_index=True)
print(df_3)

df_4 = pd.concat([df_1, df_2], axis=1)
print(df_4)

df_5 = df_3[df_3['A']>5]
print(df_5)

df_6 = df_3[df_3['A']<5]
print(df_5)


df_4 = pd.concat([df_5, df_6], axis=1, ignore_index=True)
print(df_4)


   A   B
0  1   4
1  2   5
2  3   6
3  7  10
4  8  11
5  9  12
   A  B  A   B
0  1  4  7  10
1  2  5  8  11
2  3  6  9  12
   A   B
3  7  10
4  8  11
5  9  12
   A   B
3  7  10
4  8  11
5  9  12
     0     1    2    3
3  7.0  10.0  NaN  NaN
4  8.0  11.0  NaN  NaN
5  9.0  12.0  NaN  NaN
0  NaN   NaN  1.0  4.0
1  NaN   NaN  2.0  5.0
2  NaN   NaN  3.0  6.0


**Which of the following is the correct syntax for getting the houses where no. of bedrooms is maximum having 3 views along with 1 waterfront?**

In [85]:
house_ds_1 = house_ds[(house_ds['bedrooms'] == house_ds['bedrooms'].max()) & (house_ds['waterfront'] == 1) & (house_ds['view'] == 3)]
house_ds_1

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,grade_new


In [ ]:
house_ds.loc[house_ds['bedrooms']]

16) Create a separate data frame that satisfies the conditions below. 1. Houses built before 1980 2. Have more than 2 bedrooms 3. Have more than 2 floors.

In [91]:
house_ds_1 = house_ds[(house_ds['yr_built']<1980) & (house_ds['bedrooms']>2) & (house_ds['floors']>2)]

In [89]:
house_ds

,id,date,price,bedrooms,bathrooms,sqft_living,sqft_lot,floors,waterfront,view,...,sqft_above,sqft_basement,yr_built,yr_renovated,zipcode,lat,long,sqft_living15,sqft_lot15,grade_new
0,7129300520,20141013T000000,221900,3,1.00,1180,5650,1.0,0,0,...,1180,0,1955,0,98178,47.5112,-122.257,1340,5650,9
1,6414100192,20141209T000000,538000,3,2.25,2570,7242,2.0,0,0,...,2170,400,1951,1991,98125,47.7210,-122.319,1690,7639,9
2,5631500400,20150225T000000,180000,2,1.00,770,10000,1.0,0,0,...,770,0,1933,0,98028,47.7379,-122.233,2720,8062,8
3,2487200875,20141209T000000,604000,4,3.00,1960,5000,1.0,0,0,...,1050,910,1965,0,98136,47.5208,-122.393,1360,5000,9
4,1954400510,20150218T000000,510000,3,2.00,1680,8080,1.0,0,0,...,1680,0,1987,0,98074,47.6168,-122.045,1800,7503,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
21608,263000018,20140521T000000,360000,3,2.50,1530,1131,3.0,0,0,...,1530,0,2009,0,98103,47.6993,-122.346,1530,1509,10
21609,6600060120,20150223T000000,400000,4,2.50,2310,5813,2.0,0,0,...,2310,0,2014,0,98146,47.5107,-122.362,1830,7200,10
21610,1523300141,20140623T000000,402101,2,0.75,1020,1350,2.0,0,0,...,1020,0,2009,0,98144,47.5944,-122.299,1020,2007,9
21611,291310100,20150116T000000,400000,3,2.50,1600,2388,2.0,0,0,...,1600,0,2004,0,98027,47.5345,-122.069,1410,1287,10


**17) For a given nested list, convert the same into a dataframe. sample_list = [['Carl', 22], ['Martha', 25], ['Calvin', 12], ['Stuart', 15] ] The resulting dataframe must contain the column names as ‘Name’, and ‘Age’ with the respective values from the sample_list.**

In [92]:
sample_list = [['Carl', 22], ['Martha', 25], ['Calvin', 12], ['Stuart', 15] ]
df_1 = pd.DataFrame(sample_list, columns=['Name', 'Age'])
df_1

,Name,Age
0,Carl,22
1,Martha,25
2,Calvin,12
3,Stuart,15


For a given dictionary, convert the same into a dataframe. sample_dict = {'Cristiano': ['Ronaldo','Man U', 801], 'Lionel': ['Messi','PSG', 758], 'Luis': ['Suarez','Atletico Madrid', 509], 'Robert': ['Lewandowski','Bayern Munich', 527], 'Zlatan': ['Ibrahimovic','AC Milan',553] }

In [95]:
sample_dict= {'Cristiano': ['Ronaldo','Man U', 801],
              'Lionel': ['Messi','PSG', 758],
              'Luis': ['Suarez','Atletico Madrid', 509],
              'Robert': ['Lewandowski','Bayern Munich', 527],
              'Zlatan': ['Ibrahimovic','AC Milan',553] }
sample_dict

{'Cristiano': ['Ronaldo', 'Man U', 801],
 'Lionel': ['Messi', 'PSG', 758],
 'Luis': ['Suarez', 'Atletico Madrid', 509],
 'Robert': ['Lewandowski', 'Bayern Munich', 527],
 'Zlatan': ['Ibrahimovic', 'AC Milan', 553]}

In [108]:
# df_football= pd.DataFrame(sample_dict.keys(), columns=['First Name'])
# df_football = pd.DataFrame(sample_dict.values(), columns=['Last Name', 'Club', 'Goals'])
# df_football.insert(0, 'First Name', sample_dict.keys())
# df_football
df_football = pd.DataFrame(sample_dict)
df_football = df_football.transpose()
df_football.reset_index(inplace = True)
df_football.columns = ['First Name','Last Name','Club','Goals']
df_football

,First Name,Last Name,Club,Goals
0,Cristiano,Ronaldo,Man U,801
1,Lionel,Messi,PSG,758
2,Luis,Suarez,Atletico Madrid,509
3,Robert,Lewandowski,Bayern Munich,527
4,Zlatan,Ibrahimovic,AC Milan,553
